### PR TITLE
Fix L2 metrics tests

### PR DESCRIPTION
**Pull request by oribon**
Getting a lot of flakes in the CI:
```
      Expected
          <*errors.errorString | 0xc000695790>: {
              s: "metric metallb_layer2_requests_received not found",
          }
      to be nil
```
It is caused by the fact that this metric represents the amount
of ARP/NS that arrived to the speaker for the svc's ip, while the
test doesn't necessarily does one. That can happen because the speaker
sends a gratuitous message when he announces an IP.

Also added a check for the gratuitous_sent metric to cover that.

Signed-off-by: Ori Braunshtein <obraunsh@redhat.com>

### CODE REVIEW

File: e2etest/pkg/mac/mac.go [hunk 1]
```diff
@@ -105,3 +117,69 @@ func UpdateNodeCache(nodes []corev1.Node, exec executor.Executor) error {
...
+		path, err := exec.LookPath("arping")
+		fmt.Println("ORI1 ", path, err)
+		if err != nil {
+			return err
+		}
+		cmd = path
+		args = []string{"-c", "1", "-I", iface, netIP.String()}
+	} else {
+		path, err := exec.LookPath("ndisc6")
+		fmt.Println("ORI2 ", path, err)
```
Comment (fedepaol) [inline e2etest/pkg/mac/mac.go:146]: oops

File: e2etest/pkg/mac/mac.go [hunk 2]
```diff
@@ -105,3 +117,69 @@ func UpdateNodeCache(nodes []corev1.Node, exec executor.Executor) error {
...
+	}
+
+	iface, err := ifaceForIPNetwork(netIP, exc)
+	if err != nil {
+		return err
+	}
+
+	if netIP.To4() != nil {
+		path, err := exec.LookPath("arping")
+		fmt.Println("ORI1 ", path, err)
```
Comment (fedepaol) [inline e2etest/pkg/mac/mac.go:138]: oops

File: e2etest/pkg/mac/mac.go (hunk 1 above)
Comment (oribon) [reply e2etest/pkg/mac/mac.go:146]: still under WIP trying to debug :smile:

File: e2etest/pkg/mac/mac.go [hunk 3]
```diff
@@ -105,3 +117,69 @@ func UpdateNodeCache(nodes []corev1.Node, exec executor.Executor) error {
 
 	return nil
 }
+
+// AddrResolutionReq does an ARP/NS request for the given IP.
+func AddrResolutionReq(ip string, exc executor.Executor) error {
+	var cmd string
```
Comment (fedepaol) [inline e2etest/pkg/mac/mac.go:123]: Move these below, near where you use them.

File: e2etest/pkg/mac/mac.go [hunk 4]
```diff
@@ -105,3 +117,69 @@ func UpdateNodeCache(nodes []corev1.Node, exec executor.Executor) error {
...
+		return fmt.Errorf("failed to convert %s to net.IP", ip)
+	}
+
+	iface, err := ifaceForIPNetwork(netIP, exc)
+	if err != nil {
+		return err
+	}
+
+	if netIP.To4() != nil {
+		path, err := exec.LookPath("arping")
```
Comment (fedepaol) [inline e2etest/pkg/mac/mac.go:137]: Why is this needed? If arping is in PATH, you should be able to execute it I think

File: e2etest/pkg/mac/mac.go (hunk 1 above)
Comment (oribon) [reply e2etest/pkg/mac/mac.go]: removed

File: e2etest/pkg/mac/mac.go (hunk 3 above)
Comment (oribon) [reply e2etest/pkg/mac/mac.go]: done

File: e2etest/pkg/mac/mac.go (hunk 4 above)
Comment (oribon) [reply e2etest/pkg/mac/mac.go]: Isn't needed was a leftover, removed

File: e2etest/pkg/mac/mac.go [hunk 5]
```diff
@@ -105,3 +117,69 @@ func UpdateNodeCache(nodes []corev1.Node, exec executor.Executor) error {
 
 	return nil
 }
+
+// AddrResolutionReq does an ARP/NS request for the given IP.
+func AddrResolutionReq(ip string, exc executor.Executor) error {
```
[the remark this answers is not in the document]
Comment (oribon) [reply e2etest/pkg/mac/mac.go]: done

File: e2etest/pkg/mac/mac.go [hunk 6]
```diff
@@ -105,3 +116,61 @@ func UpdateNodeCache(nodes []corev1.Node, exec executor.Executor) error {
...
+		return errors.Wrapf(err, out)
+	}
+
+	return nil
+}
+
+// ifaceForIPNetwork returns the interface name that is in the same network as the IP.
+func ifaceForIPNetwork(ip net.IP, exec executor.Executor) (string, error) {
+	ifaces := []ipIface{}
+	res, err := exec.Exec("ip", "--json", "address", "show")
```
[the remark this answers is not in the document]
Comment (oribon) [reply e2etest/pkg/mac/mac.go:153]: It returns both v4 and v6 from what I'm seeing.

[the remark this answers is not in the document]
Comment (msherif1234) [reply e2etest/pkg/mac/mac.go:153]: yep I did confirm that too Thanks

**Comment by fedepaol** [comment]
LGTM, confirmed that sudo is required for CI only an not locally.
